src/lib/comms/sol-coap.c

```diff
@@ -969,47 +969,66 @@ find_context(struct sol_coap_server *server, const struct sol_coap_resource *res
     return NULL;
 }
 
-SOL_API int
-sol_coap_notify(struct sol_coap_server *server,
-    struct sol_coap_resource *resource, struct sol_coap_packet *pkt)
+static int
+sol_coap_notify_full(struct sol_coap_server *server,
+    struct sol_coap_resource *resource, struct sol_coap_packet *pkt,
+    int (*cb)(void *cb_data, struct sol_coap_server *server,
+    struct sol_coap_resource *resource, struct sol_network_link_addr *addr,
+    struct sol_coap_packet **pkt),
+    const void *cb_data)
 {
     struct resource_observer *o;
     struct resource_context *c;
     struct sol_coap_packet *header;
     uint16_t i;
     int r = -EINVAL;
 
-    SOL_NULL_CHECK(pkt, -EINVAL);
-    SOL_NULL_CHECK_GOTO(server, error_pkt);
-    SOL_NULL_CHECK_GOTO(resource, error_pkt);
+    SOL_NULL_CHECK_GOTO(server, err_pkt);
+    SOL_NULL_CHECK_GOTO(resource, err_pkt);
 
-    COAP_RESOURCE_CHECK_API_GOTO(error_pkt);
+    COAP_RESOURCE_CHECK_API_GOTO(err_pkt);
 
     c = find_context(server, resource);
     if (!c) {
-        SOL_WRN("Context not found for expecified resource");
+        SOL_WRN("Context not found for specified resource");
         r = -ENOENT;
-        goto error_pkt;
+        goto err_pkt;
     }
 
     r = 0;
     SOL_PTR_VECTOR_FOREACH_IDX (&c->observers, o, i) {
         uint8_t type, code;
 
+        if (cb) {
+            r = cb((void *)cb_data, server, resource, &o->cliaddr, &pkt);
```

Comment:
What's the reason of creating the packet inside the callback? I.e. what does change from one observer to another?

What I had in mind is that this callback would be basically a filter, I wouldn't send a notification to that observer. Honest question: what more does it need to do?